### PR TITLE
BE: LOK-2498 : Create a timer task to re-send monitor tasks.

### PR DESCRIPTION
**Pull request by Junaid-Khan-Nant**
## Description
Create a timer that re-publishes all tasks for all tenants/locations on a regular interval.
The timer can just call MonitoredEntityService#publishTaskSet.

## Jira link(s)
https://opennms.atlassian.net/browse/LOK-2498

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/MonitorTaskTimer.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    private final MonitoredEntityService monitoredEntityService;
+    private final MonitoringLocationRepository monitoringLocationRepository;
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+    public void executeTimer(final String tenantId) {
+        scheduler.scheduleAtFixedRate(() -> republished(tenantId), 0, 1, TimeUnit.SECONDS);
+    }
+
+    private void republished(String tenantId) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/MonitorTaskTimer.java:44]:
```suggestion
    private void republish(String tenantId) {
```

**Review by fooker — CHANGES_REQUESTED** [review]
This is a little bit basic. The service should scan over **all** tenants and **all** of each tenant. The interval should be configurable.

Where is the service initialized? Can this be a spring boot service?

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/MonitorTaskTimer.java (hunk 1 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/taskset/publisher/MonitorTaskTimer.java]: Done.

**Comment by fooker** [comment]
> > This is a little bit basic. The service should scan over **all** tenants and **all** of each tenant. The interval should be configurable.
> > Where is the service initialized? Can this be a spring boot service?
> 
> @fooker, I am not sure from where it should be invoke. Do u have any suggestion about this?

I don't know, too. I guess this can be a spring service that is started on application boot.

File: inventory/main/src/main/resources/application.yml [hunk 2]
```diff
@@ -64,11 +64,15 @@ inventory:
 
 management:
   endpoints:
-    web:
```
Comment (fooker) [inline inventory/main/src/main/resources/application.yml:67]: ??

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    @EventListener(ApplicationReadyEvent.class)
+    public void executeTimer() {
+        log.info("Scheduler is calling ...");
+        scheduler.scheduleAtFixedRate(this::republish, 0, interval, TimeUnit.SECONDS);
+    }
+
+    private void republish() {
+
+        if (executed) {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:58]: This limits the whole thing to a single execution.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            log.warn("Location list is empty");
+            return;
+        }
+
+        list.stream()
+                .flatMap(location -> monitoringLocationRepository.findByTenantId(location.getTenantId()).stream())
+                .forEach(loc -> monitoredEntityService.publishTaskSet(loc.getTenantId(), loc.getId()));
+
+        executed = true;
+        stop();
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:74]: Why stop on each run?

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        try {
+            if (!scheduler.awaitTermination(10, TimeUnit.SECONDS)) {
+                scheduler.shutdownNow();
+                log.info("Scheduler stop successfully..");
+            }
+        } catch (InterruptedException e) {
+            scheduler.shutdownNow();
+            Thread.currentThread().interrupt();
+            log.warn("Scheduler stop with error..");
+        }
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:88]: Forcefully shutting down the scheduler is not required. Just drain it and let it die.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        list.stream()
+                .flatMap(location -> monitoringLocationRepository.findByTenantId(location.getTenantId()).stream())
+                .forEach(loc -> monitoredEntityService.publishTaskSet(loc.getTenantId(), loc.getId()));
+
+        executed = true;
+        stop();
+    }
+
+    public void stop() {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:77]: Need this to be hooked to an event, too.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+        var list = monitoringLocationRepository.findAll();
+        if (list.isEmpty()) {
+            log.warn("Location list is empty");
+            return;
+        }
+
+        list.stream()
+                .flatMap(location -> monitoringLocationRepository.findByTenantId(location.getTenantId()).stream())
+                .forEach(loc -> monitoredEntityService.publishTaskSet(loc.getTenantId(), loc.getId()));
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:71]: Why flat-mapping over the locations again. `list` already contains all the locations you need.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        if (executed) {
+            log.info("Already executed, skipping further execution.");
+            return;
+        }
+
+        var list = monitoringLocationRepository.findAll();
+        if (list.isEmpty()) {
+            log.warn("Location list is empty");
+            return;
+        }
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:67]: No need to warn here. Just make sure your code works on an empty list, too.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    private final MonitoringLocationRepository monitoringLocationRepository;
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+    @Value("${monitor.interval.seconds}")
+    private long interval;
+
+    private boolean executed = false;
+
+    @EventListener(ApplicationReadyEvent.class)
+    public void executeTimer() {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:51]:
```suggestion
    public void start() {
```

**Review by fooker — CHANGES_REQUESTED** [review]

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 9 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 8 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 7 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 6 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:63]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 5 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 4 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 3 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done.

File: inventory/main/src/main/resources/application.yml (hunk 2 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/resources/application.yml]: Done.

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java [hunk 10]
```diff
@@ -72,5 +74,6 @@ public void publishTaskSet(final String tenantId, final long locationId) {
                 .toList();
 
         this.taskSetPublisher.publishNewTasks(tenantId, locationId, tasts);
+        log.info("PublishTaskSet executed with tenantId :::: {}", tenantId);
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java:77]:
```suggestion
        log.info("PublishTaskSet executed with tenantId={}, locationId={}", tenantId, locationID);
```

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.ContextClosedEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class MonitorTaskTimerService {
```
Comment (fooker) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java:40]: Call it `MonitoredEntityTimer` and move it to the `monitoring` package.

**Review by fooker — CHANGES_REQUESTED** [review]

File: inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java (hunk 10 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/monitoring/MonitoredEntityService.java]: Done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java (hunk 11 above)
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/MonitorTaskTimerService.java]: Done

**Review by fooker — APPROVED** [review]

**Pull request merged**